rfc/kubernetes-operator.md

```diff
@@ -0,0 +1,89 @@
+# Kubernetes operator for Spinnaker
+
+| | |
+|-|-|
+| **Status**     | **Proposed** |
+| **RFC #**      | https://github.com/spinnaker/community/pull/41 |
+| **Author(s)**  | Nicolas Cohen (`@ncknt`) |
+| **SIG / WG**   | Kubernetes SIG |
+
+## Overview
+
+Halyard is great at offering a simple way to configure Spinnaker services but is not easy to operationalize. We've come up with an MVP of a Kubernetes operator for Spinnaker that allows administrators to manage Spinnaker with kubectl (or whatever Kubernetes tool).
+
+We're proposing to move the operator to Spinnaker's GitHub: https://docs.google.com/document/d/1Fs-xK7sOZ4YLSw1ae_cQ_rsB0kMNon9AfYli2a5Udhs/edit?usp=sharing
+
+### Goals and Non-Goals
+
+The goal of the Spinnaker Operator is to enable users to manage Spinnaker via Kubernetes native tools (`kubectl`, ...). Short term we want feature parity with Halyard deployments. Longer term, we'd like extensive preflight checks and active monitoring of Spinnaker.
+
+## Motivation and Rationale
+
+Deploying Spinnaker in Kubernetes with Halyard comes with a few challenges:
+
+- Users often start using Halyard from their laptop and struggle to find a good process to share the configuration.
+- How should Halyard itself be managed? Locally spun up at CI time or as a long-time running service?
+- How do you template the configuration if you have multiple running instances?
+- How do you pass credentials to Halyard, like kubeconfig files or AWS credentials?
+- Users are left with a lot of activities that Halyard doesn’t handle such as exposing Spinnaker and cleaning up old secrets.
+
+Finally, there’s the impression that Spinnaker is not “Kubernetes native” because it cannot be installed and managed with the usual administrator toolchain (kubectl). Talking to Spinnaker administrators, we’ve realized this is a pretty widely shared feeling. They’re already proficient at using kubectl and are reluctant to learn and configure yet another new tool.
+
+## Timeline
+
+Current MVP can deploy and expose Spinnaker. Preflight checks will be added in the coming weeks. Moving the operator to Spinnaker's GitHub account involves publishing new images to gcr and cleaning the build.
```

Comment:
What sort of preflight checks will be added (both the short-term ones and the longer-term ones) and how will they be implemented?

The implementation of preflight checks in Halyard has a number of issues, and I'd like to understand how this implementation will be different (and ensure we won't just repeat/reuse the Halyard implementation).  A couple of the issues in Halyard:
* The checks are performed using different credentials than Spinnaker will use when it's actually running
* In order to implement the checks, Halyard needs a jar-level dependency on several of the other microservices.  These microservices don't expect to be consumed as a library like this, and so very frequently break Halyard.

Reply:
We have a long list of common issues. Most of the time, it's about testing connectivity and auth. In no particular order:
- Kubernetes accounts: check connectivity, namespace, kinds, CRDs like Clouddriver does on startup. When enabled, we'd like to turn it off for CD so it can boot up faster on all pods. The check can also be done on a regular basis.
- Other provider accounts: connectivity and authorization, IAM role properly set up etc.
- Other type of checks: Version, single Spinnaker per namespace, etc.

We're working on a formalizing the framework. Happy to share that soon.

Reply:
Thanks for the details---I'd definitely like to understand the technical approach to how these pre-flight checks will be performed before signing off on this proposal.  I definitely don't want to create a situation where we create hard-to-maintain strong dependencies between the operator and implementation details of the microservices, so am very interested in the approach here.